src/Data/OrgMode/Parse/Attoparsec/Block/Markup.hs

```diff
@@ -0,0 +1,112 @@
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  Data.OrgMode.Parse.Attoparsec.Block.Markup
+-- Copyright   :  © 2014 Parnell Springmeyer
+-- License     :  All Rights Reserved
+-- Maintainer  :  Parnell Springmeyer <parnell@digitalmentat.com>
+-- Stability   :  stable
+--
+-- Parsing combinators for org-mode markup and paragraphs.
+----------------------------------------------------------------------------
+
+{-# LANGUAGE RecordWildCards   #-}
+{-# LANGUAGE OverloadedStrings   #-}
+
+module Data.OrgMode.Parse.Attoparsec.Block.Markup
+(
+  parseMarkupContent,
+  parsePlainText,
+)
+where
+
+import           Data.Semigroup                        ((<>))
+import           Data.Char                             (isSpace)
+import           Data.Text                             (Text, cons, append, cons, snoc, intercalate, dropWhileEnd, strip, stripEnd)
+import qualified Data.Text                      as     Text
+import           Data.Attoparsec.Text                  (Parser, takeWhile, choice, char, anyChar, parseOnly, isEndOfLine, endOfInput, manyTill, skipSpace)
+import           Data.OrgMode.Types                    (MarkupText (..))
+import           Prelude                        hiding (takeWhile)
+
+data Token = Token { keyChar :: Char, markup :: [MarkupText] -> MarkupText}
+
+-- | A set of token definitions for markup keywords
+tokens :: [Token]
+tokens = [ Token '*' Bold, Token '_' UnderLine, Token '/' Italic]
+
+-- | For better efficiency suggested by Attoparsec, we will hard code the token filter
+isNotToken :: Char -> Bool
+isNotToken c = c /= '*' && c /= '_' && c/= '$'
+
+-- | A Naive parser for LaTeX
+parseLaTeX :: Parser MarkupText
+parseLaTeX = char '$' *> (LaTeX <$> parseL)
+  where
+  parseL = do
+    content <- takeWhile (/= '$')
+    if Text.last content /= '\\'
+      then return content
+      else  append content <$> parseL
+
+-- | Create a markup parser based on a token
+createTokenParser :: Parser [MarkupText] -> Token -> Parser MarkupText
+createTokenParser innerParser Token{..}= do
+  -- Spaces just after the spaces
+  _ <- char keyChar <* skipSpace
+  content <- takeWhile (/= keyChar)
+  _ <- char keyChar
+  -- We need another parser passed in to parse the markup inside the markup
+  case parseOnly innerParser content of
```

Comment:
If you wanted to, you could use `bimapFold` here too:

```suggestion
  bimapFold fail (return . markup) (parseOnly innerParser content)
```

Reply:
@ixmatus May I know which package provides bimapFold?  Somehow I do not want to depend bifunctors in ghc 7.10 and ghc 8.0.1 because their compiling (with TH) takes quite long in (ghc < 8.1).